codegen/lib/graphql_swift_gen/templates/type.swift.erb

```diff
@@ -214,36 +214,41 @@ extension <%= schema_name %> {
   <% when 'INPUT_OBJECT' %>
     open class <%= type.name %> {
       <% type.required_input_fields.each do |field| %>
```

Comment:
What determines the set of required vs optional fields?

Do I understand correctly that `optional` means "field can be omitted from the mutation" rather than "field value is nullable", and that `required` means "field must be present in the mutation, whether it's nullable or not"?

If that's correct, isn't the change below this forcing all `required` fields to be non-nullable as well?

Would love some code documentation around this to clarify the implications.

Reply:
The GraphQL schema doesn't distinguish between "field can be omitted from the mutation" rather than "field value is nullable".  The name `optional` used to be more appropriate before graphql allowed a `null` input value to be specified. So `required` actually means the field is non-null in schema and `optional` that it isn't non-null in the schema.  So it is correct for a `required` field to also no be nullable.

Reply:
In GraphQL there is no distinction between "the field can be omitted from the mutation" and "the field value is nullable". They are one and the same (in the type system) so there is no way we can tell the difference. 
In practice you have fields that are implicitly required but in GraphQL are specified as nullable (not very nice right?). For example in a ProductInput the `title` property is nullable in GraphQL but if you tried to update a product with a null title we would send you a `userError`.

And yes the change below is forcing all required fields to be non nullable because if you make a mutation without a required field the mutation will always fail.

Reply:
I replied saying the same thing as Dylan 😂

Reply:
> For example in a ProductInput the title property is nullable in GraphQL but if you tried to update a product with a null title we would send you a userError.

What about update mutations that omit the `title` field altogether? I thought that was allowed, and was what we wanted, so that we wouldn't overwrite fields that we didn't modify ourselves but that might have been changed by another mutation.

Reply:
That is still possible with this PR. If you want to omit the title field all together you can do it multiple ways.
```swift 
// init with nil
let input1 = inputType(id: 1, title: nil)
print(input1.serialize) // { id: 1 }

// init with value but set serialize to false
let input2 = inputType(id: 2, title: "Great thing", serializeTitle: false)
print(input2.serialize) // { id: 2 }

// init with value but set serialize to false later
let input3 = inputType(id: 3, title: "Great thing")
input3.serializeTitle = false
print(input3.serialize) // { id: 3 }
```

Reply:
I thought `title` was a required field, and so wasn't generating `serializeTitle` at all? 😕

Reply:
Sorry in the example I was treating `title` as a non required field. If Title were a required field it would **have** to be sent to the server.

Reply:
OK, after an in-person discussion with @BenEmdon I have a better handle on the rules now:
- optional fields are nillable and expose `serialize{Field}`.

- required fields are non-nil and do not expose `serialize{Field}`.

- **fields can only be required on Input objects that are creation-only.**
 e.g. `FulfillmentInput`

- **Input objects that are used for both creation and update need all their fields marked as optional** so that we can express minimal changesets. 
e.g. `ProductInput`

- If a creation mutation omits a field which is optional, but which is implicitly required for creation, the mutation will fail with a user error.
e.g. trying to create a product with a `ProductInput` that has no `title`

- If an update mutation omits a field which is optional, but which is implicitly required for update, then the mutation will also fail with a user error (even though this is really a programmer error.)
e.g. trying to edit an existing product with a `ProductInput` that has no `id`
*(This example will be solved once we refactor update mutations to pass `id` as a separate parameter instead of making it part of the Input object.)*